comms/src/outbound_message_service/outbound_message_pool/retry_queue.rs

```diff
@@ -0,0 +1,697 @@
+        let inner = acquire_read_lock!(self.inner);
+        inner.is_empty()
+    }
+
+    /// Lease the next bucket if one is scheduled and available
+    pub fn lease_next(&self) -> Option<(K, RetryBucket<T>)> {
+        let mut inner = acquire_write_lock!(self.inner);
+        inner.take_next()
+    }
+
+    /// Return the leased bucket and make it available. Any bucket items added during the lease
+    /// are added to the returned bucket. Returns true if the lease was returned, otherwise false
+    pub fn return_lease(&self, key: &K, bucket: RetryBucket<T>) -> bool {
+        let mut inner = acquire_write_lock!(self.inner);
+        // It is an error to return a lease that doesn't exist.
+        inner.return_lease(key, bucket)
+    }
+
+    /// Add an item to a bucket. If a bucket doesn't exist for the given key one is created, otherwise
+    /// the message is appended to the available or temporary bucket.
+    pub fn add_item(&self, key: K, item: T) {
+        let mut inner = acquire_write_lock!(self.inner);
+        inner.add_item(key, item);
+    }
+
+    /// Remove a bucket from the queue. If an available/temporary bucket exists, it is returned.
+    pub fn remove(&self, key: &K) -> Option<RetryBucket<T>> {
+        let mut inner = acquire_write_lock!(self.inner);
+        inner.remove(key)
+    }
+
+    /// Clear the contents of the queue
+    pub fn clear(&self) {
+        let mut inner = acquire_write_lock!(self.inner);
+        inner.clear();
+    }
+}
+
+struct Inner<K, T>
+where K: Hash + Eq
+{
+    buckets: HashMap<K, Lease<RetryBucket<T>>>,
+    roster: BinaryHeap<ScheduledItem<K>>,
+}
+
+impl<K, T> Inner<K, T>
+where
+    K: Hash + Eq,
+    K: Clone,
+{
+    fn new() -> Self {
+        Self {
+            buckets: HashMap::new(),
+            roster: BinaryHeap::new(),
+        }
+    }
+
+    fn contains(&self, key: &K) -> bool {
+        self.buckets.contains_key(key)
+    }
+
+    fn is_empty(&self) -> bool {
+        self.buckets.is_empty()
+    }
+
+    fn clear(&mut self) {
+        self.buckets.clear();
+        self.roster.clear();
+    }
+
+    fn add_item(&mut self, key: K, item: T) {
+        match self.buckets.get_mut(&key) {
+            // Messages have already been added for this key, so we simply add it to
+            // the existing available bucket
+            Some(Lease::Available(bucket)) => {
+                bucket.push_back(item);
+            },
+            // The bucket has been taken, in the meantime other messages for the node have
+            // arrived. We fill a "taken" bucket with these messages. It is the job of
+            // the worker that has taken the original bucket to process these messages
+            // if it is able. It does this by removing the `Taken` Booking which returns
+            // the rest of the messages.
+            Some(Lease::Taken(Some(bucket))) => {
+                bucket.push_back(item);
+            },
+            Some(Lease::Taken(v @ None)) => {
```

Comment:
Huh, I haven't seen that before. Cool.

Reply:
Ye saw that in some code somewhere - first time I've used it I think :)